DRIVERS-2915 OIDC: Disallow comma character in authMechanismProperties connection string value (#1588)
mongodb/specifications: source/connection-string/connection-string-spec.md

  ```

- Key value pairs: A value that represents one or more key and value pairs. Multiple key value pairs are delimited by a
  comma (","). The key is everything up to the first colon sign (":") and the value is everything afterwards. Reserved

>   characters such as ':' must be escaped according to [RFC 2396](https://www.rfc-editor.org/rfc/rfc2396)

Reviewer: I'm not sure that we need to change anything in this spec. The second (or third...) colon must already be treated as part of the value, since the spec already says:

> up to the first colon sign and the value is everything afterwards

The original wording is problematic because it confuses restrictions on users (doc requirements) with driver requirements, and does not specify enough restricted characters, but I don't think we should correct this in this PR, but as part of future work to clarify this spec. 

Likewise, I don't think we should perform other clarifications in this PR, since the formulation is not obvious. For example, I don't think we should refer to the URI RFC, and should instead just say "percent-decoded" (we do not have URI components since connection strings are not URIs, and even if we did, we would still need to, according to the RFC, define the "set of characters actually reserved" within each particular component; note also that `%` is not in the list in section 2.2, though obviously it needs to be encoded by users).

Author: Updated

  For example:

  ```
  ?readPreferenceTags=dc:ny,rack:1
  ```

  Drivers MUST handle unencoded colon signs (":") within the value. For example:

  ```

>   ?authMechanismProperties=TOKEN_RESOURCE:mongodb%3A%2F%2F...

Reviewer: Wouldn't "unencoded" be:
```suggestion
  ?authMechanismProperties=TOKEN_RESOURCE:mongodb://...
```
But, I think this should just be a test. In general, drivers should handle all unreserved characters that are unencoded, in any context, just because the spec does not reserve them, and because percent-decoding does not alter them. This appears to be an existing consequence of how the spec is written.

Author: Clarified

  ```

  If any keys or values contain a comma (",") they MUST not be provided as part of the connection string, since it would

>   interfere with parsing.

Author: Updated

  ```

- Key value pairs: A value that represents one or more key and value pairs. Multiple key value pairs are delimited by a
  comma (","). The key is everything up to the first colon sign (":") and the value is everything afterwards.

>   For example:

Reviewer: This is a nice clarification of the example 👍

  ```
  ?readPreferenceTags=dc:ny,rack:1
  ```

  Drivers MUST handle unencoded colon signs (":") within the value. For example, given the encoded input:

  ```
  ?authMechanismProperties=TOKEN_RESOURCE%3Amongodb%3A%2F%2Ffoo
  ```

  the driver MUST interpret the key as `TOKEN_RESOURCE` and the value as `mongodb://foo`.

>   Drivers MUST document that if keys or values contain a comma (",") they MUST not be provided as part of the connection
>   string, since it would interfere with parsing.

Reviewer: ```suggestion
  For any option key-value pair that may contain a comma (such as `TOKEN_RESOURCE`), drivers MUST document that: a value containing a comma (",") MUST NOT be provided as part of the connection
  string. This prevents use of values that would interfere with parsing.
```
1. Clarifies that we do not want the "since" clause to be part of the user-facing docs.
2. Are users allowed to specify arbitrary keys? I thought we controlled them all.
3. A general comment seems like it would be confusing. We use the term "value" to refer (at least) to both: option values, and option key-value pair values. Option values can have commas, but I'm not sure how to express the distinction in user docs without using a convoluted phrase like "option key-value pair values".

It also seems fine (preferable?) to not add this new paragraph/requirement to the connection string spec at this time, since it is already quite nicely phrased in the auth spec for `TOKEN_RESOURCE` in particular.

  ```

- Key value pairs: A value that represents one or more key and value pairs. Multiple key value pairs are delimited by a

>   comma (","). The key is everything up to the first colon sign (":") and the value is everything afterwards. Drivers

Reviewer 2: >  Drivers MUST handle subsequent colon signs (":") within the value

Can we add a spec test for this?

Author: Done

>   MUST handle subsequent colon signs (":") within the value, unless otherwise specified in this document.\

Reviewer 2: > unless otherwise specified in this document

Is it otherwise specified anywhere?  If not, can we remove this part of the sentence?

Author: Removed

  If any keys
  or values contain a comma (",") they MUST not be provided as part of the connection string, since it would interfere

>   with parsing. For example:

Reviewer: ```suggestion
  with parsing. Key value pair example:
```
Since the example is not related to the preceding discussion of rules regarding `:` and `,`. (Alternatively, make the example demonstrate the rules.)

>   with parsing. For example:

Reviewer 2: With the addition of this sentence, the following example becomes confusing because the reader expects to see an example with a subsequent colon.  How about we provide two examples:

1. Keep this example, but put it directly after the sentence: `The key is everything up to the first colon sign (":") and the value is everything afterwards`
2. Add another example with a subsequent colon, e.g.

```
?authMechanismProperties=TOKEN_RESOURCE:mongodb://...
```

Author: Updated
